drivers/serial/uart_nrfx_uarte.c

```diff
@@ -0,0 +1,606 @@
+	.irq_tx_disable		= uarte_nrfx_irq_tx_disable,
+	.irq_tx_ready		= uarte_nrfx_irq_tx_ready_complete,
+	.irq_rx_enable		= uarte_nrfx_irq_rx_enable,
+	.irq_rx_disable		= uarte_nrfx_irq_rx_disable,
+	.irq_tx_complete	= uarte_nrfx_irq_tx_ready_complete,
+	.irq_rx_ready		= uarte_nrfx_irq_rx_ready,
+	.irq_err_enable		= uarte_nrfx_irq_err_enable,
+	.irq_err_disable	= uarte_nrfx_irq_err_disable,
+	.irq_is_pending		= uarte_nrfx_irq_is_pending,
+	.irq_update		= uarte_nrx_irq_update,
+	.irq_callback_set	= uarte_nrfx_irq_callback_set,
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+};
+
+static int uarte_instance_init(struct device *dev,
+			       const struct uarte_init_config *config,
+			       u8_t interrupts_active)
+{
+	int err;
+	struct device *gpio_dev;
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+
+	gpio_dev = device_get_binding(CONFIG_GPIO_NRF5_P0_DEV_NAME);
+
+	__ASSERT(gpio_dev,
+		 "UART init failed. Cannot find %s",
+		 CONFIG_GPIO_NRF5_P0_DEV_NAME);
+
+	gpio_pin_write(gpio_dev, config->pseltxd, 1);
+	gpio_pin_configure(gpio_dev,
+			   config->pseltxd,
+			   GPIO_DIR_OUT);
+	gpio_pin_configure(gpio_dev,
+			   config->pselrxd,
+			   GPIO_DIR_IN);
+
+	nrf_uarte_txrx_pins_set(uarte,
+				config->pseltxd,
+				config->pselrxd);
+
+	if (config->hwfc == NRF_UARTE_HWFC_ENABLED) {
+		gpio_pin_write(gpio_dev, config->pselrts, 1);
+		gpio_pin_configure(gpio_dev,
+				   config->pselrts,
+				   GPIO_DIR_OUT);
+		gpio_pin_configure(gpio_dev,
+				   config->pselcts,
+				   GPIO_DIR_IN);
+		nrf_uarte_hwfc_pins_set(uarte,
+					config->pselrts,
+					config->pselcts);
+	}
+
+	/* Configure flow control and parity checking */
+	nrf_uarte_configure(uarte,
+			    config->parity,
+			    config->hwfc);
+
+	err = baudrate_set(dev, config->baudrate);
+	if (err) {
+		return err;
+	}
+
+	/* Enable receiver and transmitter */
+	nrf_uarte_enable(uarte);
+
+	nrf_uarte_event_clear(uarte, NRF_UARTE_EVENT_ENDRX);
+
+	nrf_uarte_rx_buffer_set(uarte, &data->rx_data, 1);
+	nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STARTRX);
+
+#if CONFIG_UART_INTERRUPT_DRIVEN
+	if (interrupts_active) {
+
+		irq_enable(NRFX_IRQ_NUMBER_GET(uarte));
+
+		/* Set ENDTX event by requesting fake (zero-length) transfer.
+		 * Pointer to RAM variable (data->tx_buffer) is set because
+		 * otherwise such operation may result in HardFault or RAM
+		 * corruption.
+		 */
+		nrf_uarte_tx_buffer_set(uarte, data->tx_buffer, 0);
+		nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STARTTX);
+
+		/* switch off transmitter to save an energy */
+		nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STOPTX);
+	}
+#endif
+
+	return 0;
+}
+
+#define UART_NRF_UARTE_DEVICE(idx)					      \
+	DEVICE_DECLARE(uart_nrfx_uarte##idx);				      \
+	static struct uarte_nrfx_data uarte_##idx##_data;		      \
+	static const struct uarte_nrfx_config uarte_##idx##_config = {	      \
+		.uarte_regs = (NRF_UARTE_Type *)NRF_UARTE##idx		      \
```

Comment:
The register address should be coming from device tree.

Reply:
why?

Reply:
So far these values for other drivers like: UART, SPI, SPIM, SPIS, TWI, TWIS are taken from Kconfig.
I want to be consistent in this PR.

I would like to put it as it is and in other commit change approach for this and all other drivers.

Reply:
I don't have a strong opinion on this, but since this is an entirely new driver, I might be worth doing it the way we plan to have this in the future, and use DTS as much as possible. Then we also draw the guidelines for migration to DTS.

Reply:
I am on the position we shall move it to DTS, but in the second step:
https://github.com/zephyrproject-rtos/zephyr/issues/8758

Reply:
@galak @ioannisg I just spoke to @jarz-nordic and we amended the issue and added him as an assignee with a Milestone for 1.13. I think it is reasonable to do the switch all in one go so that we do things in logical steps.